src/var.c

```diff
@@ -0,0 +1,138 @@
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+
+#include <rdma/fi_errno.h>
+#include <rdma/fi_var.h>
+#include <rdma/fi_log.h>
+
+struct fi_var_t {
+	const struct fi_provider *provider;
+	char *var_name;
+	char *help_string;
+
+	char *env_var_name;
+
+	struct fi_var_t *next;
+};
+
+static struct fi_var_t *fi_vars = NULL;
+
+
+int fi_var_register(const struct fi_provider *provider, const char *var_name,
+		const char *help_string)
+{
+	int i;
+	char *str;
+	struct fi_var_t *v;
+
+	// Check for bozo cases
+	if (provider == NULL || var_name == NULL || help_string == NULL ||
+		*help_string == '\0')
+		return FI_EINVAL;
+
+	v = calloc(1, sizeof(*v));
+	if (!v)
+		return FI_ENOMEM;
+
+	v->provider = provider;
+	v->var_name = strdup(var_name);
+	v->help_string = strdup(help_string);
+	asprintf(&v->env_var_name, "FI_%s_%s",
+		v->provider->name,
+		v->var_name);
+	if (!v->var_name || !v->help_string || !v->env_var_name) {
+		free(v);
+		return FI_ENOMEM;
+	}
+
+	for (i = 0; v->env_var_name[i]; ++i)
+		v->env_var_name[i] = toupper(v->env_var_name[i]);
+
+	v->next = fi_vars;
+	fi_vars = v;
+
+	asprintf(&str, "registered var %s\n", var_name);
+	FI_INFO(provider, FI_LOG_CORE, str);
+	free(str);
+
+	return FI_SUCCESS;
+}
+
+int fi_var_get(struct fi_provider *provider, const char *var_name, char **value)
+{
+	char *str;
+	struct fi_var_t *v;
+
+	// Check for bozo cases
+	if (var_name == NULL || value == NULL)
+		return FI_EINVAL;
+
+	for (v = fi_vars; v; v = v->next)
+		if (strcmp(v->provider->name, provider->name) == 0 &&
+			strcmp(v->var_name, var_name) == 0) {
+			*value = getenv(v->env_var_name);
+
+			if (*value)
+				asprintf(&str, "read var %s=%s\n",
+					var_name, *value);
+			else
+				asprintf(&str, "read var %s=<not set>\n",
+					var_name);
+			FI_INFO(provider, FI_LOG_CORE, str);
+			free(str);
+
+			return FI_SUCCESS;
+		}
```

Comment:
Sure.  I actually like braces even for 1-line blocks (defensive programming and all that).  But I thought you kernel guys were allergic to braces, so I left them off whenever possible.  I'm happy to add them back.  :-)